### PR TITLE
Implemented BrewingStandData

### PR DESCRIPTION
**Pull request by Feeliiix**
Implementation of `SpongeBrewingStandData`.
- [x] Registration of field getters and setters
- [x] Accurately used the appropriate `AbstractData` implementation

Implementation of `ImmutableSpongeBrewingStandData`
- [x] Accurately extended the appropriate `AbstractImmutableData` implementation
- [x] Accurately instantiating final instance fields with an `ImmutableValue` counter part for any value getters
- [x] If necessary, creating a new `ImmutableValue` for a value that should not be cached, or, using the existing caching utils.

Implementation of the `RemainingBrewTimeValueProcessor`:
- [x] Appropriately extended `AbstractSpongeValueProcessor`
- [x] Individual processors for each source type possible (one for `ItemStack`/`TileEntity`/`Entity` as necessary).

Registration:
- [x] Registered the `Key` correctly by `RemainingBrewTime` in the `KeyRegistry`
- [x] Registered the `DataProcessor`s and `DataManipulatorBuilder` in `SpongeSerializationRegistry`
- [x] Registered the `ValueProcessor`s in the `SpongeSerializationRegistry`

Test Plugin code:

```java
package com.feeliiix.spongetester;

import static org.spongepowered.api.text.Texts.of;

import org.spongepowered.api.Game;
import org.spongepowered.api.block.BlockTypes;
import org.spongepowered.api.block.tileentity.carrier.BrewingStand;
import org.spongepowered.api.block.tileentity.carrier.Furnace;
import org.spongepowered.api.data.key.Keys;
import org.spongepowered.api.data.manipulator.mutable.tileentity.BrewingStandData;
import org.spongepowered.api.entity.living.player.Player;
import org.spongepowered.api.event.Listener;
import org.spongepowered.api.event.block.InteractBlockEvent;
import org.spongepowered.api.event.game.state.GameInitializationEvent;
import org.spongepowered.api.plugin.Plugin;
import org.spongepowered.api.text.Text;
import org.spongepowered.api.text.Texts;
import org.spongepowered.api.text.format.TextColors;
import org.spongepowered.api.text.format.TextStyles;
import org.spongepowered.api.util.command.CommandException;
import org.spongepowered.api.util.command.CommandResult;
import org.spongepowered.api.util.command.CommandSource;
import org.spongepowered.api.util.command.args.CommandContext;
import org.spongepowered.api.util.command.args.GenericArguments;
import org.spongepowered.api.util.command.spec.CommandExecutor;
import org.spongepowered.api.util.command.spec.CommandSpec;

@Plugin(id = "BrewingStandDataTester", name = "A plugin to test the brewing stand data", version = "1.0")
public class BrewingStandDataTest {

    private Game game; //know that i can use @Inject but my google dependency bugs

    private BrewingStandData brewingStandData;

    @Listener
    public void onInit(GameInitializationEvent event) {
        this.game = event.getGame();

        CommandSpec setCmd = CommandSpec
                .builder()
                .description(Texts.of("Set a custom data to a brewing stand"))
                .permission("spongetester.data.brewingstand.set")
                .arguments(
                        GenericArguments.onlyOne(GenericArguments.integer(of("remainingBrewTime")))
                )
                .executor(new SetCommand())
                .build();

        CommandSpec clearCmd = CommandSpec
                .builder()
                .description(Texts.of("Clear the data of a furnace"))
                .permission("spongetester.data.brewingstand.clear")
                .arguments(GenericArguments.none())
                .executor(new ClearCommand())
                .build();

        CommandSpec brewingStandDataCmd = CommandSpec
                .builder()
                .description(Texts.of("Manipulates the data of a brewing stand"))
                .permission("spongetester.data.brewingstand.*")
                .child(setCmd, "set", "s")
                .child(clearCmd, "clear", "c")
                .build();

        this.game.getCommandDispatcher().register(this, brewingStandDataCmd, "brewingStandData", "brewingStand", "bsd");
    }

    @Listener
    public void onInteract(InteractBlockEvent.Primary event) {
        if (event.getTargetBlock().getState().getType() == BlockTypes.BREWING_STAND) {
            if (brewingStandData != null) {
                event.setCancelled(true);

                BrewingStand brewingStand = (BrewingStand) event.getTargetBlock().getLocation().get().getTileEntity().get();
                brewingStand.offer(brewingStandData);

                final Text ovb = Texts.builder().append(of(TextColors.DARK_GRAY, TextStyles.STRIKETHROUGH, "                     ", TextStyles
                        .RESET, "[ ", TextColors.DARK_AQUA, " BrewingStandData ", TextColors.DARK_GRAY, "|", TextColors.AQUA, " Change Data",
                        TextColors.DARK_GRAY, " ]", TextStyles.STRIKETHROUGH, "                     ")).build();
                final Text cst = Texts.builder().append(of(TextColors.DARK_GREEN, "The data has been changed!")).build();
                final Text udb = Texts.builder().append(of(TextColors.DARK_GRAY, TextStyles.STRIKETHROUGH, "                                                                               ")).build();

                event.getCause().first(Player.class).get().sendMessage(ovb, cst, udb);
                brewingStandData = null;
            }
        }
    }

    @Listener
    public void onInteract(InteractBlockEvent.Secondary event) {
        if (event.getTargetBlock().getState().getType() == BlockTypes.BREWING_STAND) {

            final BrewingStand brewingStand = (BrewingStand) event.getTargetBlock().getLocation().get().getTileEntity().get();

            final Text ovb = Texts.builder().append(of(TextColors.DARK_GRAY, TextStyles.STRIKETHROUGH, "                       ", TextStyles.RESET,
                    "[ ", TextColors.DARK_AQUA, " BrewingStandData", TextColors.DARK_GRAY, " ]", TextStyles.STRIKETHROUGH, "                       "))
                    .build();
            final Text pbt = Texts.builder().append(of(TextColors.YELLOW, "RemainingBrewTime: ")).append(of(TextColors.GRAY, brewingStand.get(Keys
                    .REMAINING_BREW_TIME).get())).build();
            final Text udb = Texts.builder().append(of(TextColors.DARK_GRAY, TextStyles.STRIKETHROUGH, "                                                                                ")).build();

            event.getCause().first(Player.class).get().sendMessage(ovb, Texts.of(), pbt, Texts.of(), udb);
        }
    }

    private class SetCommand implements CommandExecutor {

        @Override
        public CommandResult execute(CommandSource src, CommandContext args) throws CommandException {
            if (!(src instanceof Player)) {
                src.sendMessage(Texts.builder().color(TextColors.RED)
                        .append(of("Only a player can execute this command!")).build());
                return CommandResult.empty();
            }

            final int passedBurnTime = args.<Integer>getOne("remainingBrewTime").get();
            final BrewingStandData brewingStandData = BrewingStandDataTest.this.game.getRegistry().getManipulatorRegistry().getBuilder(BrewingStandData.class)
                    .get()
                    .create();

            brewingStandData.set(Keys.REMAINING_BREW_TIME, passedBurnTime);

            BrewingStandDataTest.this.brewingStandData = brewingStandData;
            final Text ovb = Texts.builder().append(of(TextColors.DARK_GRAY, TextStyles.STRIKETHROUGH, "                   ", TextStyles.RESET, "[ "
                    + "", TextColors.DARK_AQUA, " BrewingStandData ", TextColors.DARK_GRAY, "|", TextColors.AQUA, " Change Data", TextColors
                            .DARK_GRAY,
                    " ]", TextStyles.STRIKETHROUGH, "                    ")).build();
            final Text cst = Texts.builder().append(of(TextColors.GREEN, "Left-Click a brewing stand to change its data!")).build();
            final Text udb = Texts.builder().append(of(TextColors.DARK_GRAY, TextStyles.STRIKETHROUGH, "                                                                                ")).build();
            src.sendMessage(ovb, cst, udb);

            return CommandResult.success();
        }
    }

    private class ClearCommand implements CommandExecutor {

        @Override
        public CommandResult execute(CommandSource src, CommandContext args) throws CommandException {
            if (!(src instanceof Player)) {
                src.sendMessage(Texts.builder().color(TextColors.RED)
                        .append(of("Only a player can execute this command!")).build());
                return CommandResult.empty();
            }

            final BrewingStandData brewingStandData = BrewingStandDataTest.this.game.getRegistry().getManipulatorRegistry().getBuilder(BrewingStandData.class)
                    .get()
                    .create();

            brewingStandData.set(Keys.REMAINING_BREW_TIME, 0);

            BrewingStandDataTest.this.brewingStandData = brewingStandData;
            final Text ovb = Texts.builder().append(of(TextColors.DARK_GRAY, TextStyles.STRIKETHROUGH, "                      ", TextStyles.RESET, "[ ", TextColors.DARK_AQUA, " FurnaceData ", TextColors.DARK_GRAY, "|", TextColors.AQUA, " Clear Data", TextColors.DARK_GRAY, " ]", TextStyles.STRIKETHROUGH, "                     ")).build();
            final Text cst = Texts.builder().append(of(TextColors.GREEN, "Left-Click a furnace to clear its data!")).build();
            final Text udb = Texts.builder().append(of(TextColors.DARK_GRAY, TextStyles.STRIKETHROUGH, "                                                                                ")).build();
            src.sendMessage(ovb, cst, udb);

            return CommandResult.success();
        }
    }

}
```

### CODE REVIEW

**Comment by Feeliiix** [comment]
Failed because of the newer SpongeAPI version

File: src/main/java/org/spongepowered/common/data/SpongeSerializationRegistry.java [hunk 1]
```diff
@@ -511,11 +516,14 @@ public static void setupSerialization(Game game) {
 
         final WolfWetDataProcessor wolfWetDataProcessor = new WolfWetDataProcessor();
         final ItemWetDataProcessor itemWetDataProcessor = new ItemWetDataProcessor();
-        dataRegistry.registerDataProcessorAndImpl(WetData.class, SpongeWetData.class, ImmutableWetData.class, ImmutableSpongeWetData.class, wolfWetDataProcessor);
-        dataRegistry.registerDataProcessorAndImpl(WetData.class, SpongeWetData.class, ImmutableWetData.class, ImmutableSpongeWetData.class, itemWetDataProcessor);
+        dataRegistry.registerDataProcessorAndImpl(WetData.class, SpongeWetData.class, ImmutableWetData.class, ImmutableSpongeWetData.class,
```
Comment (Aaron1011) [inline src/main/java/org/spongepowered/common/data/SpongeSerializationRegistry.java]: leave this as it is.

File: src/main/java/org/spongepowered/common/data/SpongeSerializationRegistry.java [hunk 2]
```diff
@@ -511,11 +516,14 @@ public static void setupSerialization(Game game) {
...
+                wolfWetDataProcessor);
+        dataRegistry.registerDataProcessorAndImpl(WetData.class, SpongeWetData.class, ImmutableWetData.class, ImmutableSpongeWetData.class,
+                itemWetDataProcessor);
 
         final ElderDataProcessor elderDataProcessor = new ElderDataProcessor();
-        dataRegistry.registerDataProcessorAndImpl(ElderData.class, SpongeElderData.class, ImmutableElderData.class, ImmutableSpongeElderData.class, elderDataProcessor);
+        dataRegistry.registerDataProcessorAndImpl(ElderData.class, SpongeElderData.class, ImmutableElderData.class, ImmutableSpongeElderData.class,
```
Comment (Aaron1011) [inline src/main/java/org/spongepowered/common/data/SpongeSerializationRegistry.java]: same here.

File: src/main/java/org/spongepowered/common/data/builder/block/tileentity/SpongeBrewingStandBuilder.java [hunk 3]
```diff
@@ -48,24 +48,24 @@ public SpongeBrewingStandBuilder(Game game) {
     public Optional<BrewingStand> build(DataView container) throws InvalidDataException {
```
Comment (Aaron1011) [inline src/main/java/org/spongepowered/common/data/builder/block/tileentity/SpongeBrewingStandBuilder.java]: As per @gabizou's comments, you can leave this alone - the builder system is going to be refactored.

File: src/main/java/org/spongepowered/common/data/manipulator/immutable/tileentity/ImmutableSpongeBrewingStandData.java [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+public class ImmutableSpongeBrewingStandData extends AbstractImmutableSingleData<Integer, ImmutableBrewingStandData, BrewingStandData>
+        implements ImmutableBrewingStandData {
+
+    private final ImmutableBoundedValue<Integer> immutableValue;
+
+    public ImmutableSpongeBrewingStandData(Integer value) {
+        super(ImmutableBrewingStandData.class, value, Keys.REMAINING_BREW_TIME);
+
+        this.immutableValue = SpongeValueBuilder.boundedBuilder(Keys.REMAINING_BREW_TIME)
+                .maximum(0).maximum(Integer.MAX_VALUE).defaultValue(0).actualValue(this.getValue())
```
Comment (Aaron1011) [inline src/main/java/org/spongepowered/common/data/manipulator/immutable/tileentity/ImmutableSpongeBrewingStandData.java]: The defaut value should be `400`, as that's always used in Vanilla.

File: src/main/java/org/spongepowered/common/data/manipulator/mutable/tileentity/SpongeBrewingStandData.java [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+                .compare(o.remainingBrewingTime().get(), this.getValue())
+                .result();
+    }
+
+    @Override
+    public MutableBoundedValue<Integer> remainingBrewingTime() {
+        return SpongeValueBuilder.boundedBuilder(Keys.REMAINING_BREW_TIME)
+                .minimum(0)
+                .maximum(Integer.MAX_VALUE)
+                .defaultValue(0)
```
Comment (Aaron1011) [inline src/main/java/org/spongepowered/common/data/manipulator/mutable/tileentity/SpongeBrewingStandData.java]: same here

File: src/main/java/org/spongepowered/common/data/processor/value/tileentity/MaxCookTimeValueProcessor.java [hunk 6]
```diff
@@ -54,7 +54,9 @@ public MaxCookTimeValueProcessor() {
 
     @Override
     protected boolean set(TileEntityFurnace container, Integer value) {
-        if(container.getStackInSlot(0) == null) return false; //Item cannot be null, the time depends on it
+        if (container.getStackInSlot(0) == null) {
```
Comment (Aaron1011) [inline src/main/java/org/spongepowered/common/data/processor/value/tileentity/MaxCookTimeValueProcessor.java]: keep this the same

File: src/main/java/org/spongepowered/common/data/processor/value/tileentity/PassedCookTimeValueProcessor.java [hunk 7]
```diff
@@ -54,8 +54,8 @@ public PassedCookTimeValueProcessor() {
 
     @Override
     protected boolean set(TileEntityFurnace container, Integer value) {
-        if(container.getStackInSlot(0) == null || value > container.getField(3)){ //The passedCookTime of nothing cannot be set | Cannot be higher than the maximum
-            return false;
+        if (container.getStackInSlot(0) == null || value > container.getField(3)) { //The passedCookTime of nothing cannot be set
+            return false;                                                           //Cannot be higher than the maximum
```
Comment (Aaron1011) [inline src/main/java/org/spongepowered/common/data/processor/value/tileentity/PassedCookTimeValueProcessor.java]: here too

File: src/main/java/org/spongepowered/common/data/manipulator/immutable/tileentity/ImmutableSpongeBrewingStandData.java [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+    }
+
+    @Override
+    public int compareTo(ImmutableBrewingStandData o) {
+        return ComparisonChain
+                .start()
+                .compare(o.remainingBrewTime().get(), this.getValue())
+                .result();
+    }
+}
```
Comment (Aaron1011) [inline src/main/java/org/spongepowered/common/data/manipulator/immutable/tileentity/ImmutableSpongeBrewingStandData.java]: Missing `toContainer`

**Pull request closed without merging**